### PR TITLE
t/op/magic.t - Issue #19504 - fixup tests for utf8 $0

### PR DESCRIPTION
**Pull request by demerphq**
Since PR #19334 was merged the tests for it are failing. This should fix
them. I am taking the perspective based on the broken previous tests
that that on some systems /proc/$$/commandline returns a trailing null.
This fixes the code to read /proc/$$/comandline to chop the null off if
it is there so the tests can be agnostic to that.

At the same time I have fixed the tests a bit to not use regexen where
they aren't needed and to avoid a test count issue in the skip logic,
and also to produce better errors if the code to read
/proc/$$/commandline fails.

### CODE REVIEW

File: t/op/magic.t [hunk 1]
```diff
@@ -465,16 +465,21 @@ EOP
   is ($0, $char_with_plain_pv, 'compare $0 to non-UTF8-flagged');
 
   my $linux_cmdline_cr = sub {
-    open my $rfh, '<', "/proc/$$/cmdline" or skip "open: $!", 1;
-    return do { local $/; <$rfh> };
+    my $skip= shift // 1;
```
Comment (leonerd) [inline t/op/magic.t]: Needs whitespace fix - ` = ` on both sides of `=`

File: t/op/magic.t [hunk 2]
```diff
@@ -465,16 +465,21 @@ EOP
...
+    open my $rfh, '<', "/proc/$$/cmdline"
+      or skip "failed to read '/proc/$$/cmdline': $!", $skip;
+    my $got= do { local $/; <$rfh> };
```
Comment (leonerd) [inline t/op/magic.t]: ditto

**Review by leonerd — CHANGES_REQUESTED** [review]

**Comment by demerphq** [comment]
On Mon, 7 Mar 2022 at 14:52, Paul Evans ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In t/op/magic.t
> <https://github.com/Perl/perl5/pull/19505#discussion_r820719480>:
>
> > @@ -465,16 +465,21 @@ EOP
>    is ($0, $char_with_plain_pv, 'compare $0 to non-UTF8-flagged');
>
>    my $linux_cmdline_cr = sub {
> -    open my $rfh, '<', "/proc/$$/cmdline" or skip "open: $!", 1;
> -    return do { local $/; <$rfh> };
> +    my $skip= shift // 1;
>
> Needs whitespace fix - = on both sides of =
> ------------------------------
>
> In t/op/magic.t
> <https://github.com/Perl/perl5/pull/19505#discussion_r820720747>:
>
> > @@ -465,16 +465,21 @@ EOP
>    is ($0, $char_with_plain_pv, 'compare $0 to non-UTF8-flagged');
>
>    my $linux_cmdline_cr = sub {
> -    open my $rfh, '<', "/proc/$$/cmdline" or skip "open: $!", 1;
> -    return do { local $/; <$rfh> };
> +    my $skip= shift // 1;
> +    open my $rfh, '<', "/proc/$$/cmdline"
> +      or skip "failed to read '/proc/$$/cmdline': $!", $skip;
> +    my $got= do { local $/; <$rfh> };
>
> ditto
>
>
> Those aren't bugs. I left hug my assignment operators, and since we dont
have an approved style it shouldn't matter.

But i pushed a fix to match the style in the file to get this moving.

Yves

File: t/op/magic.t [hunk 3]
```diff
@@ -465,16 +465,22 @@ EOP
   is ($0, $char_with_plain_pv, 'compare $0 to non-UTF8-flagged');
 
   my $linux_cmdline_cr = sub {
-    open my $rfh, '<', "/proc/$$/cmdline" or skip "open: $!", 1;
-    return do { local $/; <$rfh> };
+    my $skip = shift // 1;
+    open my $rfh, '<', "/proc/$$/cmdline"
+      or skip "failed to read '/proc/$$/cmdline': $!", $skip;
+    my $got = do { local $/; <$rfh> };
+    $got=~s/\0\z//;
```
Comment (leonerd) [inline t/op/magic.t:472]: FYI I'd probably write this as

```perl
return do { local $/; <$rfh> } =~ s/\0\z//r;
```

these days. It's a core perl unit test so we can rely on having 5.14+ features now. ;)

**Review by leonerd — APPROVED** [review]
LGTM

File: t/op/magic.t (hunk 3 above)
Comment (demerphq) [reply t/op/magic.t:472]: Old habits die hard. :-)

**Pull request merged**